crates/libs/windows/src/core/event.rs

```diff
@@ -0,0 +1,254 @@
+    fn len(&self) -> usize {
+        self.len
+    }
+    /// Appends a delegate to the back of the array.
+    fn push(&mut self, delegate: Delegate<T>) {
+        unsafe {
+            std::ptr::write((*self.buffer).as_mut_ptr::<Delegate<T>>().add(self.len) as _, delegate);
+            self.len += 1;
+        }
+    }
+    /// Returns a slice containing of all delegates.
+    fn as_slice(&self) -> &[Delegate<T>] {
+        if self.is_empty() {
+            &[]
+        } else {
+            unsafe { std::slice::from_raw_parts((*self.buffer).as_ptr::<Delegate<T>>() as _, self.len) }
+        }
+    }
+    /// Returns a mutable slice of all delegates.
+    fn as_mut_slice(&mut self) -> &mut [Delegate<T>] {
+        if self.is_empty() {
+            &mut []
+        } else {
+            unsafe { std::slice::from_raw_parts_mut((*self.buffer).as_mut_ptr::<Delegate<T>>() as _, self.len) }
+        }
+    }
+}
+
+impl<T: Interface + Clone> Clone for Array<T> {
+    fn clone(&self) -> Self {
+        if !self.is_empty() {
+            unsafe { (*self.buffer).0.add_ref() };
+        }
+        Self { buffer: self.buffer, len: self.len, _phantom: std::marker::PhantomData }
+    }
+}
+
+impl<T: Interface + Clone> Drop for Array<T> {
+    fn drop(&mut self) {
+        unsafe {
+            if !self.is_empty() && (*self.buffer).0.release() == 0 {
+                std::ptr::drop_in_place(self.as_mut_slice());
+                heap_free(self.buffer as _)
+            }
+        }
+    }
+}
+
+/// A reference-counted buffer.
+#[repr(C)]
+struct Buffer(RefCount);
+
+impl Buffer {
+    /// Creates a new `Buffer` with the specified size in bytes.
+    fn new(size: usize) -> Result<*mut Buffer> {
+        if size == 0 {
+            Ok(std::ptr::null_mut())
+        } else {
+            let alloc_size = std::mem::size_of::<Buffer>() + size;
+            let header = heap_alloc(alloc_size)? as *mut Buffer;
+            unsafe {
+                (*header).0 = RefCount::new(1);
+            }
+            Ok(header)
+        }
+    }
+    /// Returns a raw pointer to the buffer's contents.
+    fn as_ptr<T>(&self) -> *const T {
+        unsafe { (self as *const Self).add(1) as *const _ }
+    }
+    /// Returns a raw mutable pointer to the buffer's contents.
+    fn as_mut_ptr<T>(&mut self) -> *mut T {
+        unsafe { (self as *mut Self).add(1) as *mut _ }
+    }
+}
+
+/// Holds either a direct or indirect reference to a delegate. A direct reference is typically
+/// agile while an indirect reference is an agile wrapper.
+#[derive(Clone)]
+enum Delegate<T: Interface + Clone> {
+    Direct(T),
+    Indirect(AgileReference<T>),
+}
+
+impl<T: Interface + Clone> Delegate<T> {
+    /// Creates a new `Delegate<T>`, containing a suitable reference to the specified delegate.
+    fn new(delegate: &T) -> Self {
+        if delegate.cast::<IAgileObject>().is_err() {
+            if let Ok(delegate) = AgileReference::new(delegate) {
```

Comment:
If `delegate` does not implement `IAgileObject` but we can't create an `AgileReference` shouldn't we error?

Reply:
This is a back door that C++ uses to handle Xaml's internal delegates which refuse to support agility or marshaling, but I agree it's a hack and Rust doesn't need to support it.